### ISSUE TITLE
Unreachable child regions are leaked on freeze

### ISSUE THREAD

**Issue by plietar** [microsoft/verona#83]
```
class A { }
class B {
  f: C & iso;
}
class C { }

class Main
{
  main() {
    var x = new A;
    var y = new B in x;
    y.f = new C;
    Builtin.freeze(x);
  }
}
```

Running this gives a `debug_is_empty: found non-empty allocator` error.
The B and C objects are unreachable and should be collected on freeze. However only B is being collected.

From a quick look at `rt/region/freeze.h`, there should be a `find_iso_fields` call somewhere, probably in the `while (!to_dealloc.empty())` loop, followed by a loop to release the found regions.

**Comment by jemc** [comment]
I may have hit this same bug in my own compiler - I hit a leak in a similar situation recently and hadn't figured out how to fix it yet.

But of course I assumed that it was my codegen that was doing something wrong 😁

**Comment by mjp41** [comment]
It is slightly worse than

> From a quick look at rt/region/freeze.h, there should be a `find_iso_fields` call somewhere, probably in the `while (!to_dealloc.empty())` loop, followed by a loop to release the found regions.

The `find_iso_fields` follows fields, so you can't have deallocated any objects at this point.  It could be we find the iso fields at:

```
              p->finalise();
              to_dealloc.push(p);
```
and then deallocate the sub-isos later.  That assumes `finalisers` have read-only access to the state. Otherwise, this will require two passes of the `to_dealloc` objects.

**Comment by mjp41** [comment]
Closed by #90
